exporter/v1_compatibility.go

```diff
@@ -553,15 +705,598 @@ func makeLockMetric(m bson.M, lm lockMetric) (prometheus.Metric, error) {
+	aggregation := bson.D{
+		{Key: "$group", Value: bson.M{"_id": "$shard", "count": bson.M{"$sum": 1}}},
+	}
+
+	cursor, err := client.Database("config").Collection("chunks").Aggregate(ctx, mongo.Pipeline{aggregation})
+	if err != nil {
+		return nil, err
+	}
+
+	var shards []bson.M
+	if err = cursor.All(ctx, &shards); err != nil {
+		return nil, err
+	}
+
+	metrics := make([]prometheus.Metric, 0, len(shards))
+
+	for _, shard := range shards {
+		help := "Total number of chunks per shard"
+		labels := map[string]string{"shard": shard["_id"].(string)}
+
+		d := prometheus.NewDesc("mongodb_mongos_sharding_shard_chunks_total", help, nil, labels)
+		val, ok := shard["count"].(int32)
+		if !ok {
+			continue
+		}
+
+		metric, err := prometheus.NewConstMetric(d, prometheus.GaugeValue, float64(val))
+		if err != nil {
+			continue
+		}
+
+		metrics = append(metrics, metric)
+	}
+
+	return metrics, nil
+}
+
+func shardingShardsTotal(ctx context.Context, client *mongo.Client) (prometheus.Metric, error) {
+	n, err := client.Database("config").Collection("shards").CountDocuments(ctx, bson.M{})
+	if err != nil {
+		return nil, err
+	}
+
+	name := "mongodb_mongos_sharding_shards_total"
+	help := "Total number of shards"
+
+	d := prometheus.NewDesc(name, help, nil, nil)
+	return prometheus.NewConstMetric(d, prometheus.GaugeValue, float64(n))
+}
+
+func shardingShardsDrainingTotal(ctx context.Context, client *mongo.Client) (prometheus.Metric, error) {
+	n, err := client.Database("config").Collection("shards").CountDocuments(ctx, bson.M{"draining": true})
+	if err != nil {
+		return nil, err
+	}
+
+	name := "mongodb_mongos_sharding_shards_draining_total"
+	help := "Total number of drainingshards"
+
+	d := prometheus.NewDesc(name, help, nil, nil)
+	return prometheus.NewConstMetric(d, prometheus.GaugeValue, float64(n))
+}
+
+// ShardingChangelogSummaryID Sharding Changelog Summary ID.
+type ShardingChangelogSummaryID struct {
+	Event string `bson:"event"`
+	Note  string `bson:"note"`
+}
+
+// ShardingChangelogSummary Sharding Changelog Summary.
+type ShardingChangelogSummary struct {
+	ID    *ShardingChangelogSummaryID `bson:"_id"`
+	Count float64                     `bson:"count"`
+}
+
+// ShardingChangelogStats is an array of Sharding changelog stats.
+type ShardingChangelogStats struct {
+	Items *[]ShardingChangelogSummary
+}
+
+func changelog10m(ctx context.Context, client *mongo.Client) ([]prometheus.Metric, error) {
+	var metrics []prometheus.Metric
+
+	coll := client.Database("config").Collection("changelog")
+	match := bson.M{"time": bson.M{"$gt": time.Now().Add(-10 * time.Minute)}}
+	group := bson.M{"_id": bson.M{"event": "$what", "note": "$details.note"}, "count": bson.M{"$sum": 1}}
+
+	c, err := coll.Aggregate(ctx, []bson.M{{"$match": match}, {"$group": group}})
+	if err != nil {
+		return nil, errors.Wrap(err, ErrCannotAggregateShardingChangelogEvent.Error())
```

Comment:
So we have error ErrCannotAggregateShardingChangelogEvent that we always convert to string by calling Error(), and then immediately wrap? Why? Why ErrCannotAggregateShardingChangelogEvent is an error instead of string? Why it is even a named variable if we use it in a single place and unlikely to change?